fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -748,11 +783,109 @@ ACTOR Future<Void> commitBatch(
 							toCommit.addTag(tag);
 					}
 					toCommit.addTypedMessage(m);
-				}
-				else
+				} else if (m.type == MutationRef::Exec) {
+					state std::string param2 = m.param2.toString();
+					state ExecCmdValueString execArg(param2);
+					execArg.dbgPrint();
+					state std::string binPath = execArg.getBinaryPath();
+					state std::string uidStr = execArg.getBinaryArgValue("uid");
+
+					if (m.param1 != execDisableTLogPop
+						&& m.param1 != execEnableTLogPop
+						&& !isWhiteListed(self->whiteListedBinPathVec, binPath)) {
+						TraceEvent("ExecTransactionNotPermitted")
+							.detail("TransactionNum", transactionNum);
+						committed[transactionNum] = ConflictBatch::TransactionNotPermitted;
+					} else if (self->db->get().recoveryState != RecoveryState::FULLY_RECOVERED)  {
+						// Cluster is not fully recovered and needs TLogs
+						// from previous generation for full recovery.
+						// Currently, snapshot of old tlog generation is not
+						// supported and hence failing the snapshot request until
+						// cluster is fully_recovered.
+						TraceEvent("ExecTransactionConflict")
+							.detail("TransactionNum", transactionNum);
+						committed[transactionNum] = ConflictBatch::TransactionConflict;
+					} else {
+						// Send the ExecOp to
+						// - all the storage nodes in a single region and
+						// - only to storage nodes in local region in multi-region setup
+						// step 1: get the DatabaseConfiguration
+						state DatabaseConfiguration conf;
+						Standalone<VectorRef<KeyValueRef>> results = wait( self->txnStateStore->readRange( configKeys ) );
+						conf.fromKeyValues(results);
+						// step 2: find the tag.id from locality info of the master
+						auto localityKey =
+							self->txnStateStore->readValue(tagLocalityListKeyFor(self->master.locality.dcId())).get();
+						int8_t locality = tagLocalityInvalid;
+						if (localityKey.present()) {
+							locality = decodeTagLocalityListValue(localityKey.get());
+						}
+
+						auto ranges = self->keyInfo.intersectingRanges(allKeys);
+						std::set<Tag> allSources;
+						auto& m = (*pMutations)[mutationNum];
+						if (debugMutation("ProxyCommit", commitVersion, m))
+							TraceEvent("ProxyCommitTo", self->dbgid)
+								.detail("To", "all sources")
+								.detail("Mutation", m.toString())
+								.detail("Version", commitVersion);
+
```

Comment:
I'm stopping my review here for now, but most of this code looks an order of magnitude more expensive than I'd hope...

Iterating through every shard to get the storage servers feels like we could be iterating through and uniqifying 100,000 elements to get 500 items.

Reading and loading the entire configuration just to get the region settings seems unfortunate.  You're only using the usableRegions setting, so at minimum, you could just load that one key-value.  I'm kind of surprised that there's no variable or member of some struct laying around that already has a RegionInfo or something?

If there isn't any code for you to steal a better way of handling these things from, then when I pick the review back up, I'll try to give some more concrete advice.

Reply:
Fixed to read only the usable_regions and read all the tags from serverTagKeys.